### PR TITLE
MGMT-1755 Setup TLS trust on-prem

### PR DESCRIPTION
**Pull request by sacharya**
- Generates self-signed cert and configures nginx with the cert pair.
- Configures the agent to talk to the https endpoint with the given cacert.

### CODE REVIEW

**Comment by sacharya** [comment]
/hold

Till #182 and #185

**Comment by ronniel1** [comment]
you probably meant mgmt-1755 and not jira-1755. Please fix the commit message

File: config/onprem-iso-fcc.yaml
Comment (mhrivnak) [inline config/onprem-iso-fcc.yaml:33]: Maybe this is the right time to drop the rest of those ports. WDYT? Nothing outside of the pod should need to connect to them, except in the case of troubleshooting.

File: config/onprem-iso-fcc.yaml
Comment (rwsu) [inline config/onprem-iso-fcc.yaml:104]: IP address shouldn't be hard coded.

File: config/onprem-iso-fcc.yaml
Comment (rwsu) [inline config/onprem-iso-fcc.yaml:90]: IP address shouldn't be hard coded.

File: config/onprem-iso-fcc.yaml
Comment (rwsu) [reply config/onprem-iso-fcc.yaml:33]: +1

**Review by rwsu — CHANGES_REQUESTED** [review]

File: config/onprem-iso-fcc.yaml
Comment (sacharya) [reply config/onprem-iso-fcc.yaml]: Sounds like a good idea. Done!

File: config/onprem-iso-fcc.yaml
Comment (sacharya) [reply config/onprem-iso-fcc.yaml]: Done!

File: config/onprem-iso-fcc.yaml
Comment (sacharya) [reply config/onprem-iso-fcc.yaml:104]: This is more of a placeholder right now to indicate the dns url doesn't work until /etc/host changes.

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:123]: This looks like the changes from my branch. This should not be here.

**Review by carbonin — CHANGES_REQUESTED** [review]

File: internal/common/common.go
Comment (carbonin) [inline internal/common/common.go:20]: Also, is this the best place for a constant like this? It feels wrong here. Not sure where else it should go though. Maybe @eranco74 can suggest something?

File: internal/host/installcmd.go
Comment (carbonin) [inline internal/host/installcmd.go:98]: This is a string, right? What would this be set to? I would have expected the code below to use the value in `i.instructionConfig.ServiceCACert`.

If it's just a flag, can it be a bool called `ConfigureServiceCACert` or something?

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:361]: These errors could be more specific. Maybe `error adding CA cert to master ignition`? We may end up writing the master ignition file different times, the important part here is we couldn't set up the cert.

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:369]: Same here, mention that adding the cert is what failed.

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:487]: I don't think you need to wrap here as you have essentially the same error in the caller.

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:464]: This doesn't look like it can ever return an error, can you remove the return value?

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:181]: No need for a context if you don't have any setup and only one test, just put the condition in the test name.

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:159]: Why is this shared?

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:206]: Can you just use `""` directly rather than redefining `caCertPath`?

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:163]: Why is this shared across tests?

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:209]: Don't ignore errors. It looks like this is in a few places.

File: internal/ignition/ignition_test.go
Comment (carbonin) [inline internal/ignition/ignition_test.go:167]: If you remove the shared `err` variable I don't think you'll need this.

**Review by carbonin — CHANGES_REQUESTED** [review]
I think we should distinguish between the cert path on the service image (`/data/nginx-selfsigned.crt`) and the one that we are putting on the node (`/etc/assisted-service/service-ca-cert.crt`).
Both of those values make their way into variables called `ServiceCACert` at some point and it makes this *very* confusing to read.

Naming is hard, but I'll take a shot and suggest `ServiceCACertPath` for the one on the service and `HostCACertPath` for the one on the nodes. Try to update everywhere you use one of these (arguments and such included). This should make this much easier to understand.

Also to be clear, this PR is configuring the CA cert for the discovery ignition, controller pod, and the master/worker ignition (I still don't know a better name for this one), right? Just want to make sure since it's all in one commit.

File: internal/common/common.go
Comment (sacharya) [reply internal/common/common.go:20]: That's a good point and it did confuse me a bit. I will rename those to ServiceCACertPath and HostCACertPath.

> this PR is configuring the CA cert for the discovery ignition, controller pod, and the master/worker ignition (I still don't know a better name for this one), right? 
Yes, that's right. This PR doesn't configure SSL on the api-service itself. In the onprem usecase, we are fronting the api-service with nginx with ssl. But assuming the service is behind SSL, this configures the discovery ignition, master/worker ignition and the controller with CA Cert so they can talk to the service over SSL.

File: internal/ignition/ignition_test.go
Comment (sacharya) [reply internal/ignition/ignition_test.go]: Removed.

File: internal/ignition/ignition.go
Comment (sacharya) [reply internal/ignition/ignition.go]: Done!

File: internal/ignition/ignition.go
Comment (sacharya) [reply internal/ignition/ignition.go]: Done

File: internal/host/installcmd.go
Comment (sacharya) [reply internal/host/installcmd.go]: This is actually looking for i.instructionConfig.ServiceCACertPath. If that is set on the service, we want to have the instruction to have cacert set to HOST_CA_CERT_PATH . Hopefully, the variable renaming made it clearer.

File: internal/host/installcmd.go
Comment (carbonin) [inline internal/host/installcmd.go:98]: This is still strange to me. It looks like we never use the value in `i.instructionConfig.ServiceCACertPath`.
I guess it's just more convenient to use the existence of the `SERVICE_CA_CERT_PATH` env var in combination with `envconfig` to achieve this bool-type behavior?

File: internal/ignition/ignition.go
Comment (carbonin) [inline internal/ignition/ignition.go:58]: I feel like this is going to quickly get out of hand.

I'm not sure a parameter that is only used in some cases belongs in the generator directly. It might be worth refactoring this a bit so the user of the generator would call separate methods to make optional edits to the ignitions like adding the ca cert.

I would imagine something like this:
```
g := ignition.NewGenerator(workDir, installerDir, cluster, releaseImage, log)
g.Generate(installConfig)

if <some_condition> {
  g.AddCaCerts(serviceCACert)
}
if <some_other_condition> {
  g.SomeOtherChange(...)
}

g.UploadToS3(...)
```

File: internal/host/installcmd.go
Comment (sacharya) [reply internal/host/installcmd.go:98]: The only place where the value of ServiceCACertPath is used is to read the contents and put it into HostCACertPath, and that happens during the discovery iso ignition. The usage here is to achieve the bool-type behavior to generate the right instruction for the host. We could introduce a new env var called ConfigureServiceCACert like you said, but the presence or absense of ServiceCACertPath does the same and does it better, I think. Like there is no reason for ServiceCACertPath to be set to a value and ConfigureServiceCACert to be set to false.

File: internal/ignition/ignition.go
Comment (sacharya) [reply internal/ignition/ignition.go:58]: I see what you mean, but keep in mind that we have two implementations where the conditional logic needs to be done: job and local_job. I am also curious what it would mean with the new hostname (host level) and etc_hosts (cluster wide) ignition changes coming soon. I think we should keep it as is right now, and reevaluate with those new changes.

File: internal/host/installcmd.go
Comment (carbonin) [reply internal/host/installcmd.go:98]: Right, so it's the same env var, but we're not using the value of the `ServiceCACertPath` field of `InstructionConfig` anywhere.

I was hoping we could make the field unexported then make a method called `WriteCACert` or something but it looks like `envconfig` doesn't work with unexported fields. I'll leave it up to you if you want to create the method if you think it makes things easier to understand.

File: internal/ignition/ignition.go
Comment (carbonin) [reply internal/ignition/ignition.go:58]: > but keep in mind that we have two implementations where the conditional logic needs to be done: job and local_job

Actually that's a really good point. Won't this only ever need to be used for the on-prem case? (i.e. local job only?)

Also, obviously not for this PR, but now that this isn't actually using a kube job should we decouple this from the job construct at some point? @mhrivnak did you try this at all when moving from the kube job to local? What were the blockers (other than touching a bunch of code that we didn't want to at the time)?

File: internal/host/installcmd.go
Comment (carbonin) [reply internal/host/installcmd.go:98]: After looking at the job vs local job distinction for the other conversation I found this [flag that `main` uses](https://github.com/openshift/assisted-service/blob/0cfb84ab9a5722cc8c86efe96f2580fc433b9318/cmd/main.go#L174) to distinguish if it is running on-prem or in the kube. Maybe we should use that instead of the path in this case, it would probably also be useful for the `/etc/hosts` case.

Comment (sacharya) [reply internal/host/installcmd.go:98]: ServiceCACert is not specific to onprem only though. I expect the cloud version to use the same code as is or with minimal changes. The only difference would be how the assisted-service itself is configured with SSL (in our case, we are doing it with ngnix fronting the service), but how discovery ignition, master/work and controller flow should be the same AFAIK.

Comment (sacharya) [reply internal/host/installcmd.go]: Added `hasCACert` method.

**Review by carbonin — APPROVED** [review]
Every thing we need for the first pass looks addressed. We can handle possibly breaking out the on-prem specific ignition customizations to separate methods in the next PR or a simple follow-up.

**Pull request merged**
